src/scalar_4x64_impl.h

```diff
@@ -957,4 +957,460 @@ static SECP256K1_INLINE void secp256k1_scalar_cmov(secp256k1_scalar *r, const se
+    g[0] = (int64_t)cg & M62; cg >>= 62;
+
+    cf += (int128_t)u * f2 + (int128_t)v * g2;
+    cg += (int128_t)q * f2 + (int128_t)r * g2;
+
+    f[1] = (int64_t)cf & M62; cf >>= 62;
+    g[1] = (int64_t)cg & M62; cg >>= 62;
+
+    cf += (int128_t)u * f3 + (int128_t)v * g3;
+    cg += (int128_t)q * f3 + (int128_t)r * g3;
+
+    f[2] = (int64_t)cf & M62; cf >>= 62;
+    g[2] = (int64_t)cg & M62; cg >>= 62;
+
+    cf += (int128_t)u * f4 + (int128_t)v * g4;
+    cg += (int128_t)q * f4 + (int128_t)r * g4;
+
+    f[3] = (int64_t)cf & M62; cf >>= 62;
+    g[3] = (int64_t)cg & M62; cg >>= 62;
+
+    f[4] = (int64_t)cf;
+    g[4] = (int64_t)cg;
+}
+
+static void secp256k1_scalar_update_fg_62_var(int len, int64_t *f, int64_t *g, const int64_t *t) {
+
+    const int64_t M62 = (int64_t)(UINT64_MAX >> 2);
+    const int64_t u = t[0], v = t[1], q = t[2], r = t[3];
+    int64_t fi, gi;
+    int128_t cf, cg;
+    int i;
+
+    VERIFY_CHECK(len > 0);
+
+    fi = f[0];
+    gi = g[0];
+
+    cf = (int128_t)u * fi + (int128_t)v * gi;
+    cg = (int128_t)q * fi + (int128_t)r * gi;
+
+    VERIFY_CHECK(((int64_t)cf & M62) == 0); cf >>= 62;
+    VERIFY_CHECK(((int64_t)cg & M62) == 0); cg >>= 62;
+
+    for (i = 1; i < len; ++i) {
+
+        fi = f[i];
+        gi = g[i];
+
+        cf += (int128_t)u * fi + (int128_t)v * gi;
+        cg += (int128_t)q * fi + (int128_t)r * gi;
+
+        f[i - 1] = (int64_t)cf & M62; cf >>= 62;
+        g[i - 1] = (int64_t)cg & M62; cg >>= 62;
+    }
+
+    f[len - 1] = (int64_t)cf;
+    g[len - 1] = (int64_t)cg;
+}
+
+static void secp256k1_scalar_inverse(secp256k1_scalar *r, const secp256k1_scalar *x) {
+#if defined(EXHAUSTIVE_TEST_ORDER)
+    int i;
+    *r = 0;
+    for (i = 0; i < EXHAUSTIVE_TEST_ORDER; i++)
+        if ((i * *x) % EXHAUSTIVE_TEST_ORDER == 1)
+            *r = i;
+    /* If this VERIFY_CHECK triggers we were given a noninvertible scalar (and thus
+     * have a composite group order; fix it in exhaustive_tests.c). */
+    VERIFY_CHECK(*r != 0);
+}
+#else
+
+    /* Modular inversion based on the paper "Fast constant-time gcd computation and
+     * modular inversion" by Daniel J. Bernstein and Bo-Yin Yang. */
+
+    int64_t t[4];
+    int64_t d[5] = { 0, 0, 0, 0, 0 };
+    int64_t e[5] = { 1, 0, 0, 0, 0 };
+    int64_t f[5] = { 0x3FD25E8CD0364141LL, 0x2ABB739ABD2280EELL, 0x3FFFFFFFFFFFFFEBLL,
+        0x3FFFFFFFFFFFFFFFLL, 0xFFLL };
+    int64_t g[5];
+    secp256k1_scalar b0;
+    int i, sign;
+    uint64_t eta;
+#ifdef VERIFY
+    int zero_in = secp256k1_scalar_is_zero(x);
+#endif
+
+    b0 = *x;
+    secp256k1_scalar_encode_62(g, &b0);
+
+    /* The paper uses 'delta'; eta == -delta (a performance tweak).
+     *
+     * If the maximum bitlength of g is known to be less than 256, then eta can be set
+     * initially to -(1 + (256 - maxlen(g))), and only (741 - (256 - maxlen(g))) total
+     * divsteps are needed. */
+    eta = -(uint64_t)1;
+
+    for (i = 0; i < 12; ++i) {
```

Comment:
Is there no point in taking advantage of the fact that 741 iterations is enough (this loop looks like it supports 12*62 = 744 iterations).

@gmaxwell and I have been trying to find example inputs that need a large number of iterations (using a beam search that tries to find low bits that increase the average iteration count (over many high bits that follow) , but the best so far is 0x4212888e98e92e1c5384b37d0bf0655ec3bbad64381e713d6d6b0490d4858c83 (modulo the field order p), which needs 637 iterations.

I've also worked on some code that uses partial evaluation and interval reasoning to split up the range of inputs that have a large bound (using the log2(sqrt(f^2+4*g^2)) formula from the paper), until better bounds for the entire range is found. If my code is correct, both the field and scalar logic only need 737 iterations at most - but proving anything better will take a long time.

Reply:
For a limit of 741 I doubt there's much to be gained by saving 3 divsteps, and I prefer the simplicity. Maybe for the 32-bit version saving 9 divsteps could be more viable. Two example approaches: 1) more code (alternate divsteps method plus extra calling code) -> careful about benchmark results neglecting e.g. the instruction cache impact, or branch prediction slot. 2) pass the iteration count to divsteps -> careful not to break compiler's ability to unroll divsteps loop (or handcoded asm).

I do hope that mathematical developments could reduce the max iterations. Checking the limit for the specific prime could gain a few bits as you say. It's unclear to me whether the worst-case input(s) can be generated, but it would be wonderful if "bad" inputs could be cheaply characterised (maybe based on something like the Hamming weight in NAF), then cheaply "fixed" in constant-time (something simple like multiplication by a small constant), so that a new limit would apply.

Reply:
If we call b(x,y) = &lfloor;49/17 log<sub>2</sub>(&radic;(x<sup>2</sup> + 4y<sup>2</sup>)))&rfloor;, with appropriate corrections for small inputs, then I believe these bounds are correct:
* &delta; = 1: b(|f|,|g|) [from the paper]
* &delta; < 1: 1 - &delta; + b(|f|, ((2<sup>1-&delta;</sup>-1)|f| + |g|) / 2<sup>1-&delta;</sup>) [&delta; will only go up the next 1-&delta; steps, and in each, the g argument is either halved or averaged with f; after that many steps, this formula is an upper bound (corresponding to always averaging)]
* &delta; > 1: 1 + &delta; + b(|g|, |g| + (|f| + |g|)/2<sup>1+&delta;</sup>) [by assuming the next step will have an odd g and using the previous rule, but correcting for the fact that more &delta;-incrementing even-g steps can happen in between]

With those rules, it seems like I've proven that no input (for either field or scalar order modulus) ever needs more than 736 iterations using the code above (with finds a partition of the input range based on fixing low bits, and using partial evaluation + the bounds above to find a bound <= 736 for each).

Reply:
Update: we've proven that no inputs for either field or scalar order need more than 735 iterations. We could do 734 but I don't think it's worth the effort, as by extrapolation it would take around 16 years of CPU time (which is doable, but needs some effort to distribute the work across network nodes etc) - which means that even if proven it would make the computation hard to repeat. The &leq;736 bound takes a couple of minutes on to prove with the (somewhat involved) C++ code, and a few hours in Python.

| Bound | Proof steps for field | Proof steps for scalar |
| --- | --- | --- |
| &leq;741 | 1 | 1 |
| &leq;740 | 21 | 21 |
| &leq;739 | 85 | 85 |
| &leq;738 | 17835 | 17835 |
| &leq;737 | 6753276 | 6753276 |
| &leq;736 | 2879322241 | 2879322241 |
| &leq;735 | 1216906236685 | 1216976994782 |

Reply:
The ≤735 proof takes 80 minutes on a 128 core host.

I wouldn't mind doing the  ≤734 and wouldn't mind doing-- even without networking code-- if it would actually be even slightly useful, but it'll take me about a month realtime that way (for each field and scalar) (I could do it 4 or 5 days with a networked version).  ≤733-- by extrapolation-- would be about 7000 cpu-years and reasonably beyond my ability to do but it could be done (e.g. with the help of an actual supercomputer).

I don't think ≤732 is currently provable using all the kings cycles and all the kings mems, at least not using these techniques.

Reply:
> I don't think ≤732 is currently provable using all the kings cycles and all the kings mems, at least not using these techniques.

Surely all the kings men can muster ~1M c6g.16xlarge instances for a week (~ 350M USD).

Reply:
@sipa curious to see the code used to prove this

Reply:
@elichai https://gist.github.com/sipa/5736b83903336a1e6f3ccdeaa4cfbfea

* fgcd.cpp is a heuristic search for large inputs
* fgcd_recurse.{py,cpp} are provers for a maximum bound